Command API Implementation (Command System Overhaul)

CallumJohnson opened the pull request:
Using SubCommand.java, I have implemented AllyCommand and AddLawCommand fully using the new API, this API is very uniform and has one method callable in a loop.

Within execute, the parameter "args" is a modified array of arguments, removing the subcommand keyword means that when we handle, we act upon things after the subcommand name.
For example: 
`/f ally FactionName is treated as execute(sender, ["FactionName"], "ally");`

CallumJohnson commented:
Implementation is now at 11.76%.

CallumJohnson commented:
Implementation now at 19.61%.

CallumJohnson commented:
Implementation now at 25.49%.

dmccoystephenson reviewed (commented):
I just reviewed the SubCommand.java class and I love your implementation of the command system! Keep up the great work :) Thank you very much for your efforts in improving the plugin.

CallumJohnson commented:
Implementation at 33.33%.

CallumJohnson commented:
Implementation at 47.06%

CallumJohnson commented:
Implementation at 76.47%.

CallumJohnson commented:
Implementation at 100%.

Notes:
- Locale Messages which need to be added/removed have been tagged with a TODO (viewable in most IDEs).
- Added a Toggle for the ability to use the OLD or NEW command implementation.
- Added Deprecation to all old methodology, which could be removed once commands are tested fully.
- Added methods to Gate, and Faction class to extend functionality.

Recommendations:
- As I have been working with the API for a long time now, implementing all of the commands has displayed that the isEnemy(String name) methods and alike (isAlly, isLiege etc. etc.), could be modified or overloaded to accept a Faction object, if you provide it with a Faction object, you can still do all of the same checks, but it would be much quicker for a developer to use the API as it would read *player faction is enemy target faction*.
- I have also made a lot of changes to the way some commands work with the requirement of being a player or not, some commands, allowed players with Administrator permissions to access content, I have therefore modified this to allow both Console and Players to access this content, such as "Force" commands etc. etc. I'd like to note that if you don't need a player for an action, you can just use console to expand usability.

dmccoystephenson commented:
Exciting! I truly think these additions will vastly improve the plugin. Thank you very much for your efforts. As soon as I have time, likely this weekend, I will test all of the commands and make sure that everything still works. 

It's great that you made more commands available to use in the console. I'm sure admins using the plugin will be grateful. Your recommendations sound good, too. I'm happy to extend functionality to make things easier for developers.

dmccoystephenson commented:
I plan on reviewing and testing these changes this weekend.

dmccoystephenson commented on src/main/java/dansplugins/factionsystem/commands/HelpCommand.java:
```diff
@@ -1,13 +1,66 @@
+        super(new String[] {
+                "help", LOCALE_PREFIX + "CmdHelp"
+        }, false);
+        helpPages.put(1, Arrays.asList("Help", "List", "Info", "Members", "Join", "Leave", "Create", "Invite", "Desc"));
```
This was a cool solution!

dmccoystephenson commented on src/main/java/dansplugins/factionsystem/commands/JoinCommand.java:
```diff
@@ -11,8 +12,63 @@
+     */
+    @Override
+    public void execute(CommandSender sender, String[] args, String key) {
+
```
Are there plans for these empty methods? If, not perhaps a comment should be put stating that implementation of the method is optional or unnecessary, or maybe the empty methods should be removed. I don't mind if they stay in the implementation, though, honestly.

dmccoystephenson commented on src/main/java/dansplugins/factionsystem/commands/abs/SubCommand.java:
```diff
@@ -0,0 +1,311 @@
+     * @param args of the command.
+     * @param key of the sub-command.
+     */
+    public void performCommand(CommandSender sender, String[] args, String key) {
```
I think this method is excellent! I've always felt like doing something like this.

dmccoystephenson commented on src/main/java/dansplugins/factionsystem/data/PersistentData.java:
```diff
@@ -12,15 +12,13 @@
-    public ArrayList<PlayerPowerRecord> playerPowerRecords = new ArrayList<>();
-    public ArrayList<PlayerActivityRecord> playerActivityRecords = new ArrayList<>();
-    public ArrayList<LockedBlock> lockedBlocks = new ArrayList<>();
+    public final ArrayList<Faction> factions = new ArrayList<>();
```
Should these be final, since they may be altered during server operation?

dmccoystephenson commented on src/main/java/dansplugins/factionsystem/objects/Faction.java:
```diff
@@ -636,4 +636,28 @@ public void setPrefix(String newPrefix) {
+     * @param oldName of the Faction (dependent).
+     * @param newName of the Faction (dependent).
+     */
+    public void updateData(String oldName, String newName) {
```
This is a great idea!

dmccoystephenson reviewed (commented):
Overall, I love the changes. I'm still going to test them just to make sure everything still works, but I just wanted to look through the changes first.

dmccoystephenson reviewed (changes requested):
I did a little bit of testing and when trying to execute the info, list and help commands there is an internal error due to a null pointer exception.  It also seemed that addlaw wasn't recognized and autoclaim gave the usage message for addlaw. I'm happy to give you access to the backend of my test server for debugging purposes, otherwise I can try to debug it when I have time :)

dmccoystephenson commented:
Here is the error message that shows up in the console when I attempt to use the command '/mf list':

> [23-05-2021 04:39:30] DanTheTechMan issued server command: /f list
[23-05-2021 04:39:30] null
[23-05-2021 04:39:30] org.bukkit.command.CommandException: Unhandled exception executing command 'f' in plugin MedievalFactions v4.1-beta-2
[23-05-2021 04:39:30] at org.bukkit.command.PluginCommand.execute(PluginCommand.java:47) ~[spigot.jar:git-Spigot-758abbe-ca19c2f]
[23-05-2021 04:39:30] at org.bukkit.command.SimpleCommandMap.dispatch(SimpleCommandMap.java:149) ~[spigot.jar:git-Spigot-758abbe-ca19c2f]
[23-05-2021 04:39:30] at org.bukkit.craftbukkit.v1_16_R1.CraftServer.dispatchCommand(CraftServer.java:747) ~[spigot.jar:git-Spigot-758abbe-ca19c2f]
[23-05-2021 04:39:30] at net.minecraft.server.v1_16_R1.PlayerConnection.handleCommand(PlayerConnection.java:1702) ~[spigot.jar:git-Spigot-758abbe-ca19c2f]
[23-05-2021 04:39:30] at net.minecraft.server.v1_16_R1.PlayerConnection.a(PlayerConnection.java:1545) ~[spigot.jar:git-Spigot-758abbe-ca19c2f]
[23-05-2021 04:39:30] at net.minecraft.server.v1_16_R1.PacketPlayInChat.a(PacketPlayInChat.java:47) ~[spigot.jar:git-Spigot-758abbe-ca19c2f]
[23-05-2021 04:39:30] at net.minecraft.server.v1_16_R1.PacketPlayInChat.a(PacketPlayInChat.java:1) ~[spigot.jar:git-Spigot-758abbe-ca19c2f]
[23-05-2021 04:39:30] at net.minecraft.server.v1_16_R1.PlayerConnectionUtils.lambda$0(PlayerConnectionUtils.java:19) ~[spigot.jar:git-Spigot-758abbe-ca19c2f]
[23-05-2021 04:39:30] at net.minecraft.server.v1_16_R1.TickTask.run(SourceFile:18) ~[spigot.jar:git-Spigot-758abbe-ca19c2f]
[23-05-2021 04:39:30] at net.minecraft.server.v1_16_R1.IAsyncTaskHandler.executeTask(SourceFile:144) ~[spigot.jar:git-Spigot-758abbe-ca19c2f]
[23-05-2021 04:39:30] at net.minecraft.server.v1_16_R1.IAsyncTaskHandlerReentrant.executeTask(SourceFile:23) ~[spigot.jar:git-Spigot-758abbe-ca19c2f]
[23-05-2021 04:39:30] at net.minecraft.server.v1_16_R1.IAsyncTaskHandler.executeNext(SourceFile:118) ~[spigot.jar:git-Spigot-758abbe-ca19c2f]
[23-05-2021 04:39:30] at net.minecraft.server.v1_16_R1.MinecraftServer.aZ(MinecraftServer.java:941) ~[spigot.jar:git-Spigot-758abbe-ca19c2f]
[23-05-2021 04:39:30] at net.minecraft.server.v1_16_R1.MinecraftServer.executeNext(MinecraftServer.java:934) ~[spigot.jar:git-Spigot-758abbe-ca19c2f]
[23-05-2021 04:39:30] at net.minecraft.server.v1_16_R1.IAsyncTaskHandler.awaitTasks(SourceFile:127) ~[spigot.jar:git-Spigot-758abbe-ca19c2f]
[23-05-2021 04:39:30] at net.minecraft.server.v1_16_R1.MinecraftServer.sleepForTick(MinecraftServer.java:918) ~[spigot.jar:git-Spigot-758abbe-ca19c2f]
[23-05-2021 04:39:30] at net.minecraft.server.v1_16_R1.MinecraftServer.v(MinecraftServer.java:850) ~[spigot.jar:git-Spigot-758abbe-ca19c2f]
[23-05-2021 04:39:30] at net.minecraft.server.v1_16_R1.MinecraftServer.lambda$0(MinecraftServer.java:164) ~[spigot.jar:git-Spigot-758abbe-ca19c2f]
[23-05-2021 04:39:30] at java.lang.Thread.run(Thread.java:748) [?:1.8.0_212]
[23-05-2021 04:39:30] Caused by: java.lang.NullPointerException
[23-05-2021 04:39:30] at dansplugins.factionsystem.commands.ListCommand.execute(ListCommand.java:61) ~[?:?]
[23-05-2021 04:39:30] at dansplugins.factionsystem.commands.abs.SubCommand.performCommand(SubCommand.java:142) ~[?:?]
[23-05-2021 04:39:30] at dansplugins.factionsystem.CommandInterpreter.interpretCommand(CommandInterpreter.java:418) ~[?:?]
[23-05-2021 04:39:30] at dansplugins.factionsystem.MedievalFactions.onCommand(MedievalFactions.java:73) ~[?:?]
[23-05-2021 04:39:30] at org.bukkit.command.PluginCommand.execute(PluginCommand.java:45) ~[spigot.jar:git-Spigot-758abbe-ca19c2f]
[23-05-2021 04:39:30] ... 18 more

dmccoystephenson commented:
Here is the error message that shows up in the console when I attempt to use the command '/mf info':

> [23-05-2021 04:41:22] DanTheTechMan issued server command: /f info
[23-05-2021 04:41:22] null
[23-05-2021 04:41:22] org.bukkit.command.CommandException: Unhandled exception executing command 'f' in plugin MedievalFactions v4.1-beta-2
[23-05-2021 04:41:22] at org.bukkit.command.PluginCommand.execute(PluginCommand.java:47) ~[spigot.jar:git-Spigot-758abbe-ca19c2f]
[23-05-2021 04:41:22] at org.bukkit.command.SimpleCommandMap.dispatch(SimpleCommandMap.java:149) ~[spigot.jar:git-Spigot-758abbe-ca19c2f]
[23-05-2021 04:41:22] at org.bukkit.craftbukkit.v1_16_R1.CraftServer.dispatchCommand(CraftServer.java:747) ~[spigot.jar:git-Spigot-758abbe-ca19c2f]
[23-05-2021 04:41:22] at net.minecraft.server.v1_16_R1.PlayerConnection.handleCommand(PlayerConnection.java:1702) ~[spigot.jar:git-Spigot-758abbe-ca19c2f]
[23-05-2021 04:41:22] at net.minecraft.server.v1_16_R1.PlayerConnection.a(PlayerConnection.java:1545) ~[spigot.jar:git-Spigot-758abbe-ca19c2f]
[23-05-2021 04:41:22] at net.minecraft.server.v1_16_R1.PacketPlayInChat.a(PacketPlayInChat.java:47) ~[spigot.jar:git-Spigot-758abbe-ca19c2f]
[23-05-2021 04:41:22] at net.minecraft.server.v1_16_R1.PacketPlayInChat.a(PacketPlayInChat.java:1) ~[spigot.jar:git-Spigot-758abbe-ca19c2f]
[23-05-2021 04:41:22] at net.minecraft.server.v1_16_R1.PlayerConnectionUtils.lambda$0(PlayerConnectionUtils.java:19) ~[spigot.jar:git-Spigot-758abbe-ca19c2f]
[23-05-2021 04:41:22] at net.minecraft.server.v1_16_R1.TickTask.run(SourceFile:18) ~[spigot.jar:git-Spigot-758abbe-ca19c2f]
[23-05-2021 04:41:22] at net.minecraft.server.v1_16_R1.IAsyncTaskHandler.executeTask(SourceFile:144) ~[spigot.jar:git-Spigot-758abbe-ca19c2f]
[23-05-2021 04:41:22] at net.minecraft.server.v1_16_R1.IAsyncTaskHandlerReentrant.executeTask(SourceFile:23) ~[spigot.jar:git-Spigot-758abbe-ca19c2f]
[23-05-2021 04:41:22] at net.minecraft.server.v1_16_R1.IAsyncTaskHandler.executeNext(SourceFile:118) ~[spigot.jar:git-Spigot-758abbe-ca19c2f]
[23-05-2021 04:41:22] at net.minecraft.server.v1_16_R1.MinecraftServer.aZ(MinecraftServer.java:941) ~[spigot.jar:git-Spigot-758abbe-ca19c2f]
[23-05-2021 04:41:22] at net.minecraft.server.v1_16_R1.MinecraftServer.executeNext(MinecraftServer.java:934) ~[spigot.jar:git-Spigot-758abbe-ca19c2f]
[23-05-2021 04:41:22] at net.minecraft.server.v1_16_R1.IAsyncTaskHandler.awaitTasks(SourceFile:127) ~[spigot.jar:git-Spigot-758abbe-ca19c2f]
[23-05-2021 04:41:22] at net.minecraft.server.v1_16_R1.MinecraftServer.sleepForTick(MinecraftServer.java:918) ~[spigot.jar:git-Spigot-758abbe-ca19c2f]
[23-05-2021 04:41:22] at net.minecraft.server.v1_16_R1.MinecraftServer.v(MinecraftServer.java:850) ~[spigot.jar:git-Spigot-758abbe-ca19c2f]
[23-05-2021 04:41:22] at net.minecraft.server.v1_16_R1.MinecraftServer.lambda$0(MinecraftServer.java:164) ~[spigot.jar:git-Spigot-758abbe-ca19c2f]
[23-05-2021 04:41:22] at java.lang.Thread.run(Thread.java:748) [?:1.8.0_212]
[23-05-2021 04:41:22] Caused by: java.lang.NullPointerException
[23-05-2021 04:41:22] at dansplugins.factionsystem.commands.InfoCommand.execute(InfoCommand.java:59) ~[?:?]
[23-05-2021 04:41:22] at dansplugins.factionsystem.commands.abs.SubCommand.performCommand(SubCommand.java:142) ~[?:?]
[23-05-2021 04:41:22] at dansplugins.factionsystem.CommandInterpreter.interpretCommand(CommandInterpreter.java:418) ~[?:?]
[23-05-2021 04:41:22] at dansplugins.factionsystem.MedievalFactions.onCommand(MedievalFactions.java:73) ~[?:?]
[23-05-2021 04:41:22] at org.bukkit.command.PluginCommand.execute(PluginCommand.java:45) ~[spigot.jar:git-Spigot-758abbe-ca19c2f]
[23-05-2021 04:41:22] ... 18 more

dmccoystephenson commented:
I tried removing the 'final' keyword from the data structures in PersistantData.java but the above errors weren't fixed. Still, I recommend making this change.

CallumJohnson commented:
I have removed the "final" keyword as per your request.
I have fixed and tested fixes applied to a surplus of commands.

I did notice that the original implementation of the /f join (faction-name) never checked if the player was invited, this has been fixed in the newest implementation.

I also noticed that the /mf config command (without arguments), returns this message: 
"[key 'ValidSubCommandsShowSet' not found]"
Which I doubt is desired behaviour, I triple checked my definitions by copying and pasting the key from your own code, with the same output.

When teleporting to the home of the Faction, once the player has been teleported, a message could be added saying "Teleported" etc, etc.

dmccoystephenson reviewed (changes requested):
I've tested most of the commands and I've put together a spreadsheet to document the results. I've messaged you the link.

Most of the commands function as expected, but there are 12 that have problems ranging from incorrectly colored messages to an internal error.

I don't mind fixing the incorrectly colored messages after merging, but would you mind fixing the other problems when you have the chance?

dmccoystephenson reviewed (changes requested):
A lot of the problems have been fixed! There remains just two major problems now:
- For many of the force sub-commands, the usage message shows up even when the command is used correctly. This is true for the peace, demote, join, kick, power, transfer and removevassal sub-commands.
- Even when using the invoke command correctly, the "Allied faction and warring faction must be designated inside single quotes!" shows up.

Would you mind attempting to fix the remaining two major problems above?

Thanks again for all your hard work 👍 

(There are also some minor problems with some other commands, but these problems are either also in v4.0 or so minor that they should be very easy to fix for the v4.1 update. I don't mind taking a look at them later.)

CallumJohnson commented on src/main/java/dansplugins/factionsystem/utils/ArgumentParser.java:
```diff
@@ -31,7 +31,7 @@ public String createStringFromFirstArgOnwards(String[] args) {
         ArrayList<String> toReturn = new ArrayList<>();
 
-        String argumentString = ArgumentParser.getInstance().createStringFromFirstArgOnwards(args);
+        String argumentString = String.join(" ", args);
```
I just wanna mention that I did triple check the usage of this method and the only times it is used is by methods which I have now modified to use the new command implementation, making the "crecreateStringFromFirstArgOnwards" method obsolete and I remove the unrequired argument from the String array.

- Callum

dmccoystephenson replied:
Sounds good!

dmccoystephenson reviewed (changes requested):
Many of the force commands work, now, and the invoke command has given me no trouble.

I did encounter three more problems, though, one of which is fairly critical:
- You are able to create a faction that has the same name as an existing faction, currently.
- The '/mf force join' command works with some players and not others.
- The '/mf force kick' command shows the "Player not found!" message even if player is online

I'm fine with saving the latter two to fix later, but would you mind looking into the first problem?

dmccoystephenson commented on src/main/java/dansplugins/factionsystem/commands/CreateCommand.java:
```diff
@@ -11,8 +12,58 @@
+        if (args.length == 0) {
+            player.sendMessage(translate("&c" + getText("UsageCreate")));
+            return;
+        }
```
It should be fairly simple to fix the problem mentioned above by adding the following code segment to the checks that happen before the creation:


```
if (PersistentData.getInstance().getFaction(name) != null) {
    player.sendMessage(ChatColor.RED + LocaleManager.getInstance().getText("FactionAlreadyExists"));
    return false;
}
```

If you want, I can even do it after merging the pull request.

dmccoystephenson replied:
I've decided to handle this afterwards.

dmccoystephenson reviewed (approved):
After testing once more, I can comfortably merge this pull request. Thank you very much for your contribution to the project!
